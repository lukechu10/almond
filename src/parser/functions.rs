//! Parsing for JS functions and programs.

use crate::ast::*;
use crate::parser::util::*;
use crate::parser::*;
use nom::InputLength;
use nom_locate::position;

/// Parses a complete JS program
pub fn parse_program(s: Span) -> ParseResult<Node> {
    let (s, start) = position(s)?;
    let (s, _) = sp0(s)?; // eat all preceding whitespace

    let (s, body) = parse_function_body_inner(s)?;
    if s.input_len() != 0 {
        panic!("The source code was not completely parsed. This is a bug with the parser.\nUnparsed code starting on line {}", s.location_line());
    }

    let (s, end) = position(s)?; // Program loc should include all trailing whitespace
    Ok((s, NodeKind::Program { body }.with_pos(start, end)))
}

pub fn parse_declaration(s: Span) -> ParseResult<Node> {
    parse_function_declaration(s)
}

pub fn parse_function_declaration(s: Span) -> ParseResult<Node> {
    let parse_function_declaration_signature = pair(
        separated_pair(
            ws0(opt(keyword_async)),
            ws1(keyword_function),
            parse_identifier,
        ),
        delimited(ws0(tag("(")), parse_formal_param_list, ws0(tag(")"))),
    );

    context(
        "function declaration",
        map(
            spanned(pair(
                parse_function_declaration_signature,
                parse_function_body,
            )),
            |((((is_async, id), params), body), start, end)| {
                NodeKind::FunctionDeclaration {
                    function: Function {
                        id: Box::new(Some(id)),
                        params,
                        body: Box::new(body),
                        is_async: is_async.is_some(),
                    },
                }
                .with_pos(start, end)
            },
        ),
    )(s)
}

pub fn parse_function_expr(s: Span) -> ParseResult<Node> {
    let parse_function_expr_signature = pair(
        separated_pair(
            ws0(opt(keyword_async)),
            ws0(keyword_function),
            opt(parse_identifier),
        ),
        delimited(ws0(tag("(")), parse_formal_param_list, ws0(tag(")"))),
    );

    context(
        "function expression",
        map(
            spanned(pair(parse_function_expr_signature, parse_function_body)),
            |((((is_async, id), params), body), start, end)| {
                NodeKind::FunctionExpression {
                    function: Function {
                        id: Box::new(id),
                        params,
                        body: Box::new(body),
                        is_async: is_async.is_some(),
                    },
                }
                .with_pos(start, end)
            },
        ),
    )(s)
}

pub fn parse_function_body(s: Span) -> ParseResult<Node> {
    context(
        "function body",
        map(
            spanned(delimited(
                ws0(tag("{")),
                parse_function_body_inner,
                ws0(tag("}")),
            )),
            |(body, start, end)| NodeKind::BlockStatement { body }.with_pos(start, end),
        ),
    )(s)
}

pub fn parse_function_body_inner(s: Span) -> ParseResult<Vec<Node>> {
    let parse_directive_list = many0(parse_directive);
    let parse_source_element_list = many0(parse_source_elem);

    context(
        "function body inner",
        map(
            pair(parse_directive_list, parse_source_element_list),
            |(mut directives, stmts)| {
                directives.extend(stmts);
                directives
            },
        ),
    )(s)
}

pub fn parse_formal_param_list(s: Span) -> ParseResult<Vec<Node>> {
    terminated(
        separated_list0(ws0(char(',')), parse_formal_param),
        // trailing comma
        opt(ws0(char(','))),
    )(s)
}

pub fn parse_formal_param(s: Span) -> ParseResult<Node> {
    parse_identifier(s)
}

/// Example: `"use strict";`
pub fn parse_directive(s: Span) -> ParseResult<Node> {
    let parse_directive_literal = spanned(terminated(ws0(string_lit), opt(ws0(semi))));
    map(parse_directive_literal, |(node, start, end)| {
        match &node.kind {
            NodeKind::Literal { value } => match value {
                LiteralValue::String(directive) => {
                    let directive = directive.clone();
                    NodeKind::ExpressionStatement {
                        expression: Box::new(node),
                        directive: Some(directive),
                    }
                    .with_pos(start, end)
                }
                _ => unreachable!(
                    "string_lit should return a literal with value LiteralValue::String"
                ),
            },
            _ => unreachable!("string_lit should return a Node::Identifier"),
        }
    })(s)
}

pub fn parse_source_elem(s: Span) -> ParseResult<Node> {
    alt((parse_declaration, parse_stmt))(s)
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_json_snapshot;

    #[test]
    fn test_directive() {
        assert_json_snapshot!(parse_directive("\"use strict\"".into()).unwrap().1);
    }

    #[test]
    fn test_function_declaration() {
        assert_json_snapshot!(
            parse_function_declaration(
                r#"function a() {
                    "use strict"
                    return 1
                }"#
                .into()
            )
            .unwrap()
            .1
        );
        assert_json_snapshot!(
            parse_function_declaration(r#"function a() {}"#.into())
                .unwrap()
                .1
        );
        assert_json_snapshot!(
            parse_function_declaration(r#"function a(param1, param2) {}"#.into())
                .unwrap()
                .1
        );
        assert_json_snapshot!(
            parse_function_declaration(r#"function a(param1, param2,) {}"#.into())
                .unwrap()
                .1
        );
        assert_json_snapshot!(
            parse_function_declaration(
                r#"function a() {
                    "use strict"
                    "also directive"
                    a;
                    "not directive"
                    return 1
                }"#
                .into()
            )
            .unwrap()
            .1
        );
        assert_json_snapshot!(
            parse_function_declaration(
                r#"function fib() {
                    if (x === 0) return 0;
                    else if (x === 1) return 1;
                    else return fib(x - 1) + fib(x - 2);
                }"#
                .into()
            )
            .unwrap()
            .1
        );
        assert_json_snapshot!(
            parse_program(
                r#"function foo() {
                    // stuff
                }; foo();"# // note the `;` after the function declaration. Should be parsed as an `EmptyStatement`.
                    .into()
            )
            .unwrap()
            .1
        );
    }

    #[test]
    fn test_nested_function_declaration() {
        assert_json_snapshot!(
            parse_function_declaration(
                r#"function a() {
                    function b() {
                        a();
                    }
                }"#
                .into()
            )
            .unwrap()
            .1
        );
    }

    #[test]
    fn test_function_expr() {
        parse_function_declaration(
            r#"function a() {
                // function expressions should be wrapped up in parenthesis
                function () {
                    a();
                }
            }"#
            .into(),
        )
        .unwrap_err();

        assert_json_snapshot!(
            parse_function_declaration(
                r#"function a() {
                    // anonymous function:
                    (function () {
                        a();
                    })
                }"#
                .into()
            )
            .unwrap()
            .1
        );

        assert_json_snapshot!(
            parse_function_declaration(
                r#"function a() {
                    // function expression
                    0, function b() {}
                }"#
                .into()
            )
            .unwrap()
            .1
        );
    }

    #[test]
    fn parse_immediately_invoked_function_expr() {
        assert_json_snapshot!(
            parse_stmt(
                r#"(function () {
                    123; // something
                })()"#
                    .into()
            )
            .unwrap()
            .1
        );
    }

    #[test]
    fn test_program() {
        assert_json_snapshot!(
            parse_program(
                r#"var x = 1;
                function foo() {
                    return x;
                }"#
                .into()
            )
            .unwrap()
            .1
        );
        assert_json_snapshot!(
            parse_program(
                r#"var x = 1;
                function foo() {
                    if (x) {
                        return x;
                    }
                }"#
                .into()
            )
            .unwrap()
            .1
        );
        assert_json_snapshot!(
            parse_program(
                r#"
                function fib() {
                    if (x === 0) return 0;
                    else if (x === 1) return 1;
                    else return fib(x - 1) + fib(x - 2);
                }
                var x = fib(10);"#
                    .into()
            )
            .unwrap()
            .1
        );
        assert_json_snapshot!(
            parse_program(
                r#"function foo() {
                    test;
                    // abc "123"-
                    // test
                    foo();
                }"#
                .into()
            )
            .unwrap()
            .1
        );
    }

    #[test]
    fn test_empty_function() {
        assert_json_snapshot!(
            parse_program(
                r#"function foo(callback) {
                    // A👍a
                }"#
                .into()
            )
            .unwrap()
            .1
        );
    }

    #[test]
    fn parse_program_immediately_invoked_function_expr() {
        assert_json_snapshot!(
            parse_program(
                r#"(function () {
                    123; // something
                })()"#
                    .into()
            )
            .unwrap()
            .1
        );
        assert_json_snapshot!(
            parse_program(
                r#"(function (global, factory) {
                    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports) :
                        typeof define === 'function' && define.amd ? define(['exports'], factory) :
                            (global = global || self, factory(global.Library = {}));
                }(this, (function (exports) {})));"#
                    .into()
            )
            .unwrap()
            .1
        );
    }

    #[test]
    fn test_async_function() {
        assert_json_snapshot!(
            parse_program(
                r#"async function foo(callback) {
                    await test();
                }"#
                .into()
            )
            .unwrap()
            .1
        );
    }
}
