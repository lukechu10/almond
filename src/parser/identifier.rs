//! Parsing for JS identifiers

use crate::ast::*;
use crate::parser::util::*;
use crate::parser::*;
use nom_locate::position;

/// Parses any 4 hex digits unicode escape sequence.
pub fn unicode_esc_seq(s: Span) -> ParseResult<char> {
    let (s, hex_str) = recognize(count(one_of("1234567890abcdefABCDEF"), 4))(s)?;
    let hex_u32 = u32::from_str_radix(*hex_str, 16).unwrap(); // FIXME
    let char = std::char::from_u32(hex_u32).unwrap(); // FIXME
    Ok((s, char))
}

/// Parses `"\\u"` followed by any 4 hex digits unicode escape sequence.
pub fn identifier_unicode_esc_sequence(s: Span) -> ParseResult<char> {
    preceded(tag("\\u"), unicode_esc_seq)(s)
}

pub fn identifier_start(s: Span) -> ParseResult<char> {
    verify(alt((identifier_unicode_esc_sequence, anychar)), |c: &char| match c {
        // These are already matched by is_xid_start but are explicitly at the beginning for performance reasons.
        'a'..='z' => true,
        'A'..='Z' => true,
        c if unicode_xid::UnicodeXID::is_xid_start(*c) => true,
        '$' | '_' => true,
        _ => false,
    })(s)
}

pub fn identifier_continue(s: Span) -> ParseResult<char> {
    verify(alt((identifier_unicode_esc_sequence, anychar)), |c: &char| match c {
        // These are already matched by is_xid_continue but are explicitly at the beginning for performance reasons.
        'a'..='z' => true,
        'A'..='Z' => true,
        c if unicode_xid::UnicodeXID::is_xid_continue(*c) => true,
        '$' | '_' => true,
        _ => false,
    })(s)
}

/// ```ebnf
/// IdentifierName
/// ```
pub fn parse_identifier_name(s: Span) -> ParseResult<Node> {
    context("identifier", |s| {let (s, start) = position(s)?;
    let (s, name) =
        map(
            pair(identifier_start, many0(identifier_continue)),
            |(c, v)| format!("{}{}", c, v.into_iter().collect::<String>()),
        )(s)?;
    let (s, end) = position(s)?;

    let (s, _) = sp0(s)?;
    Ok((s, NodeKind::Identifier { name }.with_pos(start, end)))})(s)
}

/// ```ebnf
/// Identifier :: IdentifierName but not ReservedWord
/// ```
pub fn parse_identifier(s: Span) -> ParseResult<Node> {
    context("identifier", |s| {let (s, start) = position(s)?;
    let (s, name) = verify(
        map(
            pair(identifier_start, many0(identifier_continue)),
            |(c, v)| format!("{}{}", c, v.into_iter().collect::<String>()),
        ),
        |ident: &str| 
            // make sure identifier is not a reserved word
            // not(reserved_word2)(ident.into()).is_ok(),
            not(reserved_word)(ident.into()).is_ok()
    )(s)?;
    let (s, end) = position(s)?;

    let (s, _) = sp0(s)?;
    Ok((s, NodeKind::Identifier { name }.with_pos(start, end)))})(s)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_identifier() {
        all_consuming(parse_identifier)("myVar".into()).unwrap();
        all_consuming(parse_identifier)("abc123".into()).unwrap();
        all_consuming(parse_identifier)("$".into()).unwrap();
        all_consuming(parse_identifier)("_".into()).unwrap();
        all_consuming(parse_identifier)("$elem".into()).unwrap();
        all_consuming(parse_identifier)("_elem".into()).unwrap();
        all_consuming(parse_identifier)("$123".into()).unwrap();
        all_consuming(parse_identifier)("_123".into()).unwrap();

        // identifiers can start with keywords
        all_consuming(parse_identifier)("var_id".into()).unwrap();
        all_consuming(parse_identifier)("class_id".into()).unwrap();
    }

    #[test]
    fn test_identifier_with_trailing_whitespace() {
        all_consuming(parse_identifier)("abc123 ".into()).unwrap();
        all_consuming(parse_identifier)("abc123 \t\n ".into()).unwrap();
    }

    #[test]
    fn test_bad_identifier() {
        all_consuming(parse_identifier)("123abc".into()).unwrap_err();
    }

    #[test]
    fn test_keyword_not_identifier() {
        all_consuming(parse_identifier)("var".into()).unwrap_err();
        all_consuming(parse_identifier)("class".into()).unwrap_err();
        all_consuming(parse_identifier)("true".into()).unwrap_err();
        all_consuming(parse_identifier)("null".into()).unwrap_err();
    }

    #[test]
    fn test_identifier_with_unicode_escape() {
        assert_eq!(
            parse_identifier(r#"abc\u0061"#.into()).unwrap().1.kind,
            NodeKind::Identifier {
                name: "abca".into()
            }
        );
        assert_eq!(
            parse_identifier(r#"abc\u0061\u0062123"#.into())
                .unwrap()
                .1
                .kind,
            NodeKind::Identifier {
                name: "abcab123".into()
            }
        );
    }
}
