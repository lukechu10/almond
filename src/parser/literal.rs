//! Parsing for JS literals.

use crate::ast::*;
use crate::parser::util::*;
use crate::parser::*;

pub fn parse_literal(s: Span) -> ParseResult<Node> {
    ws0(alt((
        null_lit,
        bool_lit,
        numeric_lit,
        string_lit,
        array_lit,
        object_lit,
        regex_lit, // defined in regex.rs
    )))(s)
}

pub fn null_lit(s: Span) -> ParseResult<Node> {
    map(
        spanned(pair(tag("null"), not(identifier_continue))),
        |(_, start, end)| LiteralValue::Null.into_node_kind().with_pos(start, end),
    )(s)
}

pub fn bool_lit(s: Span) -> ParseResult<Node> {
    map(
        spanned(alt((
            value(false, pair(tag("false"), not(identifier_continue))),
            value(true, pair(tag("true"), not(identifier_continue))),
        ))),
        |(val, start, end)| {
            LiteralValue::Boolean(val)
                .into_node_kind()
                .with_pos(start, end)
        },
    )(s)
}

pub fn numeric_lit(s: Span) -> ParseResult<Node> {
    map(
        spanned(alt((octal_int_lit, hex_int_lit, decimal_lit))),
        |(num, start, end)| {
            LiteralValue::Number(num)
                .into_node_kind()
                .with_pos(start, end)
        },
    )(s)
}

fn decimal_lit(s: Span) -> ParseResult<f64> {
    recognize_float(s).map(|(s, float_str)| (s, float_str.parse::<f64>().unwrap()))
}

fn octal_int_lit(s: Span) -> ParseResult<f64> {
    map(preceded(char('0'), oct_digit1), |octal_str: Span| {
        i64::from_str_radix(octal_str.fragment(), 8).unwrap() as f64
    })(s)
}

fn hex_int_lit(s: Span) -> ParseResult<f64> {
    map(
        preceded(alt((tag("0x"), tag("0X"))), hex_digit1),
        |hex_str: Span| i64::from_str_radix(hex_str.fragment(), 16).unwrap() as f64,
    )(s)
}

/// Parses a valid character in a double quote string.
fn character_double_quote(s: Span) -> ParseResult<char> {
    let (input, c) = none_of("\"")(s)?;
    if c == '\\' {
        alt((
            preceded(char('u'), unicode_esc_seq),
            map_res(anychar, |c| {
                Ok(match c {
                    '\'' | '"' | '\\' | '/' => c,
                    'b' => '\x08',
                    'f' => '\x0C',
                    'n' => '\n',
                    'r' => '\r',
                    't' => '\t',
                    // vertical tab
                    'v' => '\x0b',
                    '0' => '\0',
                    _ => c,
                    #[allow(unreachable_patterns)]
                    // Only for type inference purposes. Case on top should catch all cases.
                    _ => return Err(()),
                })
            }),
        ))(input)
    } else {
        Ok((input, c))
    }
}

/// Parses a valid character in a single quote string.
fn character_single_quote(s: Span) -> ParseResult<char> {
    let (input, c) = none_of("\'")(s)?;
    if c == '\\' {
        alt((
            preceded(char('u'), unicode_esc_seq),
            map_res(anychar, |c| {
                Ok(match c {
                    '\'' | '"' | '\\' | '/' => c,
                    'b' => '\x08',
                    'f' => '\x0C',
                    'n' => '\n',
                    'r' => '\r',
                    't' => '\t',
                    // vertical tab
                    'v' => '\x0b',
                    '0' => '\0',
                    _ => c,
                    #[allow(unreachable_patterns)]
                    // Only for type inference purposes. Case on top should catch all cases.
                    _ => return Err(()),
                })
            }),
        ))(input)
    } else {
        Ok((input, c))
    }
}

pub fn string_lit(s: Span) -> ParseResult<Node> {
    context(
        "string literal",
        map(
            spanned(alt((
                context(
                    "string literal double quotes",
                    delimited(
                        char('"'),
                        fold_many0(character_double_quote, String::new(), |mut string, c| {
                            string.push(c);
                            string
                        }),
                        char('"'),
                    ),
                ),
                context(
                    "string literal single quotes",
                    delimited(
                        char('\''),
                        fold_many0(character_single_quote, String::new(), |mut string, c| {
                            string.push(c);
                            string
                        }),
                        char('\''),
                    ),
                ),
            ))),
            |(string, start, end)| {
                LiteralValue::String(string)
                    .into_node_kind()
                    .with_pos(start, end)
            },
        ),
    )(s)
}

pub fn array_lit(s: Span) -> ParseResult<Node> {
    map(
        spanned(delimited(
            ws0(char('[')),
            parse_expr_list_with_opt_expr,
            ws0(char(']')),
        )),
        |(expr_list, start, end)| {
            NodeKind::ArrayExpression {
                elements: expr_list,
            }
            .with_pos(start, end)
        },
    )(s)
}

fn parse_property_name(s: Span) -> ParseResult<Node> {
    alt((
        parse_identifier_name, // not parse_identifier
        ws0(string_lit),
        ws0(numeric_lit),
    ))(s)
}

fn parse_property_assignment(s: Span) -> ParseResult<Node> {
    let simple = map(
        spanned(separated_pair(
            parse_property_name,
            ws0(tag(":")),
            parse_expr_no_seq,
        )),
        |((key, value), start, end)| {
            NodeKind::Property {
                key: Box::new(key),
                value: Box::new(value),
                kind: PropertyKind::Init,
            }
            .with_pos(start, end)
        },
    );
    let getter = map(
        spanned(tuple((
            ws0(keyword_get),
            parse_property_name,
            spanned(tuple((ws0(tag("(")), ws0(tag(")")), parse_function_body))),
        ))),
        |((_, key, ((_, _, body), start_func, end_func)), start, end)| {
            NodeKind::Property {
                key: Box::new(key),
                value: Box::new(
                    NodeKind::FunctionExpression {
                        function: Function {
                            body: Box::new(body),
                            id: Box::new(None),
                            params: Vec::new(),
                            is_async: false, // async is incompatible with get
                        },
                    }
                    .with_pos(start_func, end_func),
                ),
                kind: PropertyKind::Get,
            }
            .with_pos(start, end)
        },
    );
    let setter = map(
        spanned(tuple((
            ws0(keyword_set),
            parse_property_name,
            spanned(tuple((
                ws0(tag("(")),
                parse_formal_param,
                ws0(tag(")")),
                parse_function_body,
            ))),
        ))),
        |((_, key, ((_, param, _, body), start_func, end_func)), start, end)| {
            NodeKind::Property {
                key: Box::new(key),
                value: Box::new(
                    NodeKind::FunctionExpression {
                        function: Function {
                            body: Box::new(body),
                            id: Box::new(None),
                            params: vec![param],
                            is_async: false, // async is incompatible with set
                        },
                    }
                    .with_pos(start_func, end_func),
                ),
                kind: PropertyKind::Set,
            }
            .with_pos(start, end)
        },
    );

    alt((getter, setter, simple))(s)
}

pub fn object_lit(s: Span) -> ParseResult<Node> {
    map(
        spanned(delimited(
            ws0(tag("{")),
            alt((
                // with trailing comma
                terminated(
                    separated_list1(ws0(tag(",")), parse_property_assignment),
                    ws0(tag(",")),
                ),
                // without trailing comma
                separated_list0(ws0(tag(",")), parse_property_assignment),
            )),
            ws0(tag("}")),
        )),
        |(properties, start, end)| NodeKind::ObjectExpression { properties }.with_pos(start, end),
    )(s)
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_json_snapshot;

    #[test]
    fn test_bool_lit() {
        bool_lit("true".into()).unwrap();
        bool_lit("false".into()).unwrap();
    }

    #[test]
    fn test_bool_lit_with_trailing_whitespace() {
        bool_lit("true ".into()).unwrap();
        bool_lit("false ".into()).unwrap();
    }

    #[test]
    fn test_bad_bool_lit() {
        bool_lit("bad".into()).unwrap_err();
        bool_lit("badfalse".into()).unwrap_err();
        bool_lit("falsebad".into()).unwrap_err();
    }

    #[test]
    fn test_numeric_lit() {
        assert_eq!(
            numeric_lit("123".into()).unwrap().1.kind,
            LiteralValue::Number(123.0).into_node_kind()
        );
        assert_eq!(
            numeric_lit("0123".into()).unwrap().1.kind,
            LiteralValue::Number(83.0).into_node_kind() // in octal
        );
        assert_eq!(
            numeric_lit("0x123".into()).unwrap().1.kind,
            LiteralValue::Number(291.0).into_node_kind() // in hex
        );
        assert_eq!(
            numeric_lit("0".into()).unwrap().1.kind,
            LiteralValue::Number(0.0).into_node_kind()
        );
        assert_eq!(
            numeric_lit("0x0".into()).unwrap().1.kind,
            LiteralValue::Number(0.0).into_node_kind() // hex 0
        );
    }

    #[test]
    fn test_string_lit() {
        assert_eq!(
            parse_literal(r#""my string""#.into()).unwrap().1.kind,
            LiteralValue::String("my string".into()).into_node_kind()
        );
        assert_eq!(
            parse_literal(r#"'my string'"#.into()).unwrap().1.kind,
            LiteralValue::String("my string".into()).into_node_kind()
        );
        assert_eq!(
            parse_literal(r#"'"quoted"'"#.into()).unwrap().1.kind,
            LiteralValue::String("\"quoted\"".into()).into_node_kind()
        );
        assert_eq!(
            parse_literal(r#""'single'""#.into()).unwrap().1.kind,
            LiteralValue::String("'single'".into()).into_node_kind()
        );
        assert_eq!(
            parse_literal(r#""\"""#.into()).unwrap().1.kind,
            LiteralValue::String("\"".into()).into_node_kind()
        );
        assert_eq!(
            parse_literal(r#""\n""#.into()).unwrap().1.kind,
            LiteralValue::String("\n".into()).into_node_kind()
        );
        assert_eq!(
            parse_literal(r#""?|\\\\[^\\r\\n\\f]|[\\w-]|[^\0-\\x7f])+""#.into())
                .unwrap()
                .1
                .kind,
            LiteralValue::String("?|\\\\[^\\r\\n\\f]|[\\w-]|[^\0-\\x7f])+".into()).into_node_kind()
        );
        // unicode escape sequences
        assert_eq!(
            parse_literal(r#""\uFFFD""#.into()).unwrap().1.kind,
            LiteralValue::String("\u{fffd}".into()).into_node_kind()
        );
        assert_eq!(
            parse_literal(r#""\ufffd""#.into()).unwrap().1.kind,
            LiteralValue::String("\u{fffd}".into()).into_node_kind()
        );
    }

    #[test]
    fn test_array_lit() {
        array_lit("[]".into()).unwrap();
        array_lit("[true, true]".into()).unwrap();
    }

    #[test]
    fn test_array_lit_with_whitespace() {
        array_lit("[ true, true ]\n".into()).unwrap();
        array_lit("[   true,\n true\t ]\t".into()).unwrap();
    }

    #[test]
    fn test_array_lit_with_trailing_comma() {
        assert_json_snapshot!(array_lit("[ true, true, ]\n".into()).unwrap().1);
        array_lit("[   true,\n true,\t ]\t".into()).unwrap();
    }

    #[test]
    fn test_obj_lit() {
        assert_json_snapshot!(parse_literal(r#"{ abc: "foo" }"#.into()).unwrap().1);
        parse_literal(r#"{ abc: 1, 2, 3 }"#.into()).unwrap_err();
        assert_json_snapshot!(
            parse_literal(r#"{ abc: 123, def: { foo: "bar" } }"#.into())
                .unwrap()
                .1
        );
        assert_json_snapshot!(parse_literal(r#"{}"#.into()).unwrap().1);
        assert_json_snapshot!(parse_literal(r#"{ }"#.into()).unwrap().1);

        assert_json_snapshot!(
            parse_literal(
                r#"{
                    test: function () {
                        return {};
                    },
                    value: 1 + 2,
                    foo: "abc"
                }"#
                .into()
            )
            .unwrap()
            .1
        );
        assert_json_snapshot!(
            parse_literal(
                r#"{
                    get: function () {
                        return {};
                    }
                }"#
                .into()
            )
            .unwrap()
            .1
        );
        assert_json_snapshot!(
            parse_literal(
                r#"{
                    get: test() ? function () {
                        return 0;
                    } : function () {
                        return 1;
                    },
                    foo: bar
                }"#
                .into()
            )
            .unwrap()
            .1
        );
    }

    #[test]
    fn test_obj_lit_with_trailing_comma() {
        assert_json_snapshot!(parse_literal(r#"{ abc: "foo", }"#.into()).unwrap().1);
        assert_json_snapshot!(
            parse_literal(r#"{ abc: 123, def: { foo: "bar", }, }"#.into())
                .unwrap()
                .1
        );
        parse_literal(r#"{,}"#.into()).unwrap_err();
        parse_literal(r#"{ , }"#.into()).unwrap_err();
    }

    #[test]
    fn test_obj_lit_getter_setter() {
        assert_json_snapshot!(
            parse_literal(
                r#"{
                    get a() {
                        a;
                    },
                    set b(x) {
                        b;
                    }
                }"#
                .into()
            )
            .unwrap()
            .1
        );
    }
}
