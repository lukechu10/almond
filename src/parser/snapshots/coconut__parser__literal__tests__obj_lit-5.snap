---
source: src/parser/literal.rs
expression: "parse_literal(r#\"{\n                    test: function () {\n                        return {};\n                    },\n                    value: 1 + 2,\n                    foo: \"abc\"\n                }\"#.into()).unwrap().1"
---
{
  "type": "ObjectExpression",
  "properties": [
    {
      "type": "Property",
      "key": {
        "type": "Identifier",
        "name": "test",
        "start": 22,
        "end": 26
      },
      "value": {
        "type": "FunctionExpression",
        "id": null,
        "params": [],
        "body": {
          "type": "BlockStatement",
          "body": [
            {
              "type": "ReturnStatement",
              "argument": {
                "type": "ObjectExpression",
                "properties": [],
                "start": 73,
                "end": 75
              },
              "start": 66,
              "end": 76
            }
          ],
          "start": 40,
          "end": 98
        },
        "async": false,
        "start": 28,
        "end": 98
      },
      "kind": "init",
      "start": 22,
      "end": 98
    },
    {
      "type": "Property",
      "key": {
        "type": "Identifier",
        "name": "value",
        "start": 120,
        "end": 125
      },
      "value": {
        "type": "BinaryExpression",
        "operator": "+",
        "left": {
          "type": "Literal",
          "value": 1.0,
          "start": 127,
          "end": 128
        },
        "right": {
          "type": "Literal",
          "value": 2.0,
          "start": 131,
          "end": 132
        },
        "start": 127,
        "end": 132
      },
      "kind": "init",
      "start": 120,
      "end": 132
    },
    {
      "type": "Property",
      "key": {
        "type": "Identifier",
        "name": "foo",
        "start": 154,
        "end": 157
      },
      "value": {
        "type": "Literal",
        "value": "abc",
        "start": 159,
        "end": 164
      },
      "kind": "init",
      "start": 154,
      "end": 164
    }
  ],
  "start": 0,
  "end": 182
}
